### PR TITLE
Change S3 storage to use the AWS default region provider chain

### PR DESCRIPTION
**Pull request by davemx**
Changes S3 storage to use the AWS default region provider chain if the region is not specified in the POM configuration. Previously, these classes used a system property of `AWS_DEFAULT_REGION`, which is not standard for AWS environments. The standard AWS client builders use the `DefaultAwsRegionProviderChain`, which allows for the standard environment variable naming, system property naming, profiles, EC2 instance config, etc.

This change will require updating the blog to remove mention of using the `AWS_DEFAULT_REGION` system property. It also mentions using it as an environment variable, but that does not currently work. So some usage would need to be updated.

If you want to continue allowing for `-DAWS_DEFAULT_REGION=region-name`, it would be more standard to create a new region provider chain, add this custom region picking strategy, but then set the rest of the chain to the default.

I hope this helps!

AWS Docs: [AWS SDK for Java - AWS Region Selection](https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/java-dg-region-selection.html)

### CODE REVIEW

**Comment by sgandon** [comment]
Looking at the doc you mentionned above, it seems that the region can be set using the environment variable : `AWS_REGION` if the region is not set in the builder.
Did you check that providing the region this way works ?
If that is so then you are absolutly right and the resolution of the region that I did add is completly useless.

**Comment by davemx** [comment]
Let me write some unit tests for it, if you don't mind. I didn't want to make too many changes since I'm not a maintainer but I'll refine it as a proper testable fix.

**Comment by davemx** [comment]
Okay, I refactored to use a more proper provider chain, which let me write decent unit tests for the `AWS_REGION` environment variable and `aws.region` system property.

**Comment by davemx** [comment]
And we can add support for `AWS_DEFAULT_REGION` if you want to maintain backwards compatibility.

**Comment by gkatzioura** [comment]
https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html

AWS_DEFAULT_REGION is the default environmental variable. For a different region configuration they can use this environmental variable or property or declare the region. (Or even through the region property).

Currently the works

- When a cli is preconfigured

- When default environmental variables are set

- When people provide the credentials from these properties

The above should continue to be supported. Is there a use case which is currently not supported and this pull request makes it possible?

File: S3StorageWagon/src/test/java/com/gkatzioura/maven/cloud/s3/S3StorageRegionProviderChainTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    @Test
+    public void testSystemPropertyRegion() {
+        System.setProperty(AWS_REGION_SYSTEM_PROPERTY, SYSTEM_PROPERTY_REGION);
+
+        final S3StorageRegionProviderChain regionProvider = new S3StorageRegionProviderChain();
+        Assert.assertEquals(SYSTEM_PROPERTY_REGION, regionProvider.getRegion());
+
+        System.clearProperty(AWS_REGION_SYSTEM_PROPERTY);
+    }
+
```
Comment (sgandon) [inline S3StorageWagon/src/test/java/com/gkatzioura/maven/cloud/s3/S3StorageRegionProviderChainTest.java]: what about a test with a null region ?

**Comment by davemx** [comment]
> Is there a use case which is currently not supported and this pull request makes it possible?

Yeah. It wasn't working with the region, mostly due to this bug:

S3StorageRepository.java, line 119: `builder.withRegion(region.get());`

It should either bet `setRegion` or `builder = builder.withRegion(...)`.

**Comment by gkatzioura** [comment]
@davemx
All good will test and merge

**Comment by gkatzioura** [comment]
will add support for the AWS_DEFAULT_REGION

**Pull request merged**